{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "c60343a0-26b8-4da0-8190-5ae457611de3",
   "metadata": {
    "scrolled": true,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(polyester)\n",
    "library(Biostrings)\n",
    "#library(devtools)\n",
    "#install_github(\"alyssafrazee/polyester\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "43b32843-f83d-4a2c-93a5-740fd16d14b7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1] ‘1.29.1’"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "packageVersion(\"polyester\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a01725ec",
   "metadata": {},
   "source": [
    "### GRCH38   \n",
    "each batch randomly select 1000 transcripts, the accoaiste counts starts from 1 reads - 500reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec678c02-5093-4e45-8b9f-759b29052801",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file <- \"../04_simulation/GRCh38.HERVK.fa\"\n",
    "fasta = readDNAStringSet(fasta_file)\n",
    "set.seed(4)\n",
    "\n",
    "n <- 100\n",
    "\n",
    "for(b in seq(1,10) ) {\n",
    "    \n",
    "    fasta_subset <- fasta[sample(length(fasta), n)]\n",
    "\n",
    "    dir_name =  paste0(\"../04_simulation/GRCh38_Batches\",b)\n",
    "    cat(\"working on dir:\", dir_name, \"\\n\")\n",
    "    if (!dir.exists(dir_name)) {\n",
    "      dir.create(dir_name, recursive = TRUE)\n",
    "    }\n",
    "\n",
    "    fa.file <- paste0(dir_name,\"/batch.\",b,\".fasta\")\n",
    "    writeXStringSet(fasta_subset, filepath = fa.file )\n",
    "    \n",
    "    random_numbers <- runif(n,min = 0, max = 1)\n",
    "    # Scale values it range [30, 300]\n",
    "    min_val <- 1\n",
    "    max_val <- 300\n",
    "    # Rescale beta values to desired range\n",
    "    readspertx <- as.integer(min_val + (max_val - min_val) * random_numbers)\n",
    "\n",
    "    ctl <- rep(1, n)\n",
    "    trt <- sample(c(0.5, 1, 2), size=n,\n",
    "        prob=c(0.05, 0.9, 0.05), replace=TRUE)\n",
    "    fc <- cbind(ctl, trt)\n",
    "\n",
    "    simulate_experiment(fa.file, reads_per_transcript=readspertx, seed = 1 ,readlen = 150,fraglen = 400,fragsd = 50 , \n",
    "             num_reps=c(2,2) , outdir= dir_name , strand_specific = TRUE, fold_changes =  fc  )\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "221200c2-34c4-4a73-a640-d693288ce7e0",
   "metadata": {},
   "source": [
    "## GRCh38 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0937977-a7d0-41ef-b544-0ecee269775e",
   "metadata": {},
   "source": [
    "## Maternal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5d55fbe",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3e5ae36b-c37e-4820-9e3f-0bbb45653cac",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# FASTA annotation\n",
    "#fasta_file = system.file('extdata', 'chr22.fa', package='polyester')\n",
    "fasta_file <- \"../04_simulation//HG002.mat.HERVK.fa\"\n",
    "fasta = readDNAStringSet(fasta_file)\n",
    "\n",
    "set.seed(123)\n",
    "n <- length(fasta)\n",
    "alpha <- 1  # Shape parameter α (controls skewness)\n",
    "beta <- 5   # Shape parameter β (controls skewness)\n",
    "# Generate beta-distributed values\n",
    "random_numbers <- rbeta(n, alpha, beta)\n",
    "# Scale values it range [30, 300]\n",
    "min_val <- 1\n",
    "max_val <- 300\n",
    "# Rescale beta values to desired range\n",
    "readspertx <- as.integer(min_val + (max_val - min_val) * random_numbers)\n",
    "\n",
    "fold_changes = rep(1 , n)\n",
    "fsum <- sum(readspertx > 30)\n",
    "fold_changes[readspertx > 30] = sample(c(0.5, 1, 2), size=fsum, prob=c(0.1, 0.8, 0.1), replace=TRUE)\n",
    "#fold_changes = sample(c(0.5, 1, 2), size=n, prob=c(0.05, 0.9, 0.05), replace=TRUE)\n",
    "#readspertx > \n",
    "\n",
    "simulate_experiment(fasta_file, reads_per_transcript=readspertx, seed = 1,readlen = 150,fraglen = 400,fragsd = 50,\n",
    "    num_reps=c(5,5), outdir='../04_simulation/simulated_reads1', fold_changes = fold_changes, strand_specific = TRUE) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85113d8a-2d63-481d-93d7-d61eb189e275",
   "metadata": {},
   "source": [
    "## paternal "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "864dd0e4-cab3-4a11-8a30-8d35e6342330",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../04_simulation/Batch1_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch2_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch3_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch4_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch5_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch6_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch7_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch8_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch9_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch10_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch11_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch12_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch13_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch14_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch15_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch16_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch17_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch18_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch19_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch20_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch21_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch22_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch23_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch24_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch25_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch26_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch27_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch28_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch29_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch30_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch31_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch32_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch33_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch34_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch35_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch36_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch37_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch38_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch39_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch40_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch41_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch42_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch43_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch44_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch45_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch46_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch47_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch48_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch49_HG002_paternalbegin simulate\n",
      "../04_simulation/Batch50_HG002_paternalbegin simulate\n"
     ]
    }
   ],
   "source": [
    "set.seed(4)\n",
    " fasta_file <- \"../04_simulation//HG002.pat.HERVK.fa\"\n",
    "fasta = readDNAStringSet(fasta_file)\n",
    "    \n",
    "n <- length(fasta)\n",
    "\n",
    "for(b in seq(1,50)){\n",
    "    dir_name = paste0(\"../04_simulation/Batch\",b,\"_HG002_paternal\")\n",
    "    cat(dir_name)\n",
    "    # FASTA annotation\n",
    "\n",
    "    random_numbers <- runif(n,min = 0, max = 1)\n",
    "    # Scale values it range [30, 300]\n",
    "    min_val <- 1\n",
    "    max_val <- 500\n",
    "    # Rescale beta values to desired range\n",
    "    readspertx <- as.integer(min_val + (max_val - min_val) * random_numbers)\n",
    "\n",
    "    fold_changes = rep(1 , n * 2)\n",
    "    fold_changes = sample(c(0.5, 1, 2), size= n * 2, prob=c(0.1, 0.8, 0.1), replace=TRUE)\n",
    "    fold_changes = matrix(fold_changes, nrow = n)\n",
    "    cat(\"begin simulate\\n\")\n",
    "    simulate_experiment(fasta_file, reads_per_transcript=readspertx, seed = 1,readlen = 150,fraglen = 400,fragsd = 50, seed=142,\n",
    "             num_reps=c(3,3), outdir= dir_name , fold_changes = fold_changes, strand_specific = TRUE) \n",
    "    \n",
    "    #simulate_experiment( fasta_file, reads_per_transcript=readspertx, seed = 1,readlen = 150,fraglen = 400,fragsd = 50,seed=142,\n",
    "     #                  meanmodel = FALSE , num_reps=c(3,3), outdir= dir_name ,\n",
    "      #                 fold_changes = fold_changes, strand_specific = TRUE,reportCoverage = TRUE ) \n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2b5a31fb-32d7-4c24-b587-582e2c0cf3f7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>2921</li><li>2</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 2921\n",
       "\\item 2\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 2921\n",
       "2. 2\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 2921    2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(fold_changes )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1404f17b-e724-4d2e-9062-7f2cb49bd050",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table style=\"width: 100%;\"><tr><td>simulate_experiment {polyester}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>simulate RNA-seq experiment using negative binomial model</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>create FASTA files containing RNA-seq reads simulated from provided\n",
       "transcripts, with optional differential expression between two groups\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>simulate_experiment(\n",
       "  fasta = NULL,\n",
       "  gtf = NULL,\n",
       "  seqpath = NULL,\n",
       "  outdir = \".\",\n",
       "  num_reps = c(10, 10),\n",
       "  reads_per_transcript = 300,\n",
       "  size = NULL,\n",
       "  fold_changes,\n",
       "  paired = TRUE,\n",
       "  reportCoverage = FALSE,\n",
       "  ...\n",
       ")\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table>\n",
       "<tr><td><code id=\"fasta\">fasta</code></td>\n",
       "<td>\n",
       "<p>path to FASTA file containing transcripts from which to simulate\n",
       "reads. See details.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"gtf\">gtf</code></td>\n",
       "<td>\n",
       "<p>path to GTF file containing transcript structures from which reads\n",
       "should be simulated. See details.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"seqpath\">seqpath</code></td>\n",
       "<td>\n",
       "<p>path to folder containing one FASTA file (<code>.fa</code>\n",
       "extension) for each chromosome in <code>gtf</code>. See details.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"outdir\">outdir</code></td>\n",
       "<td>\n",
       "<p>character, path to folder where simulated reads should be\n",
       "written, with *no* slash at the end. By default, reads are\n",
       "written to current working directory.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"num_reps\">num_reps</code></td>\n",
       "<td>\n",
       "<p>How many biological replicates should be in each group? The\n",
       "length <code>num_reps</code> determines how many groups are in the experiment.\n",
       "For example, <code>num_reps = c(5,6,5)</code> specifies a 3-group experiment with\n",
       "5 samples in group 1, 6 samples in group 2, and 5 samples in group 3.\n",
       "Defaults to a 2-group experiment with 10 reps per group (i.e.,\n",
       "<code>c(10,10)</code>).</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"reads_per_transcript\">reads_per_transcript</code></td>\n",
       "<td>\n",
       "<p>baseline mean number of reads to simulate\n",
       "from each transcript. Can be an integer, in which case this many reads\n",
       "are simulated from each transcript, or an integer vector whose length\n",
       "matches the number of transcripts in <code>fasta</code>. Default 300. You can\n",
       "also leave <code>reads_per_transcript</code> empty and set <code>meanmodel=TRUE</code>\n",
       "to draw baseline mean numbers from a model based on transcript length.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"size\">size</code></td>\n",
       "<td>\n",
       "<p>the negative binomial <code>size</code> parameter (see\n",
       "<code>NegBinomial</code>) for the number of reads drawn per transcript.\n",
       "It can be a matrix (where the user can specify the size parameter per\n",
       "transcript, per group), a vector (where the user can specify the size per\n",
       "transcript, perhaps relating to reads_per_transcript), or a single number,\n",
       "specifying the size for all transcripts and groups.\n",
       "If left NULL, defaults to <code>reads_per_transcript * fold_changes / 3</code>.\n",
       "Negative binomial variance is mean + mean^2 / size.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"fold_changes\">fold_changes</code></td>\n",
       "<td>\n",
       "<p>Matrix specifying multiplicative fold changes\n",
       "between groups. There is no default, so you must provide this argument.\n",
       "In real data sets, lowly-expressed transcripts often show high fold\n",
       "changes between groups, so this can be kept in mind when setting\n",
       "<code>fold_changes</code> and <code>reads_per_transcript</code>. This argument must\n",
       "have the same number of columns as there are groups as\n",
       "specified by <code>num_reps</code>, and must have the same number of rows as\n",
       "there are transcripts in <code>fasta</code>. A fold change of X in matrix entry\n",
       "i,j means that for replicate j, the baseline mean number of reads\n",
       "(reads_per_transcript[i]) will be multiplied by X. Note that the\n",
       "multiplication happens before the negative binomial value\n",
       "(for the number of reads that *actually will* be\n",
       "drawn from transcript i, for replicate j) is drawn. This argument is\n",
       "ignored if <code>length(num_reps)</code> is 1 (meaning you only have 1 group in\n",
       "your simulation).</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"paired\">paired</code></td>\n",
       "<td>\n",
       "<p>If <code>TRUE</code>, paired-end reads are simulated; else\n",
       "single-end reads are simulated. Default <code>TRUE</code></p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"reportCoverage\">reportCoverage</code></td>\n",
       "<td>\n",
       "<p>whether to write out coverage information to\n",
       "<code>sample_coverages.rda</code> file in the <code>outdir</code>.\n",
       "defaults to <code>FALSE</code></p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"...\">...</code></td>\n",
       "<td>\n",
       "<p>any of several other arguments that can be used to add nuance\n",
       "to the simulation. See details.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>Reads can either be simulated from a FASTA file of transcripts\n",
       "(provided with the <code>fasta</code> argument) or from a GTF file plus DNA\n",
       "sequences (provided with the <code>gtf</code> and <code>seqpath</code> arguments).\n",
       "Simulating from a GTF file and DNA sequences may be a bit slower: it took\n",
       "about 6 minutes to parse the GTF/sequence files for chromosomes 1-22, X,\n",
       "and Y in hg19.\n",
       "</p>\n",
       "<p>Several optional parameters can be passed to this function to adjust the\n",
       "simulation. The options are:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li> <p><code>readlen</code>: read length. Default 100.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>lib_sizes</code>: Library size factors for the biological replicates.\n",
       "<code>lib_sizes</code> should have length equal to the total number of\n",
       "replicates in the experiment, i.e., <code>sum(num_reps)</code>. For each\n",
       "replicate, once the number of reads to simulate from each transcript for\n",
       "that replicate is known, all read numbers across all transcripts from that\n",
       "replicate are multiplied by the corresponding entry in <code>lib_sizes</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>distr</code> One of 'normal', 'empirical', or 'custom', which\n",
       "specifies the distribution from which to draw RNA fragment lengths. If\n",
       "'normal', draw fragment lengths from a normal distribution. You can provide\n",
       "the mean of that normal distribution with <code>fraglen</code> (defaults to 250)\n",
       "and the standard deviation of that normal distribution with <code>fragsd</code>\n",
       "(defaults to 25). You can provide a single number for each, or a\n",
       "vector with length equal to the total number of samples.\n",
       "If 'empirical', draw fragment lengths\n",
       "from a fragment length distribution estimated from a real data set. If\n",
       "'custom', draw fragment lengths from a custom distribution, which you can\n",
       "provide as the <code>custdens</code> argument. <code>custdens</code> should be a\n",
       "density fitted using <code>logspline</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>error_model</code>: The error model can be one of:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li> <p><code>'uniform'</code>: errors are distributed uniformly across reads.\n",
       "You can also provide an <code>'error_rate'</code> parameter, giving the overall\n",
       "probability of making a sequencing error at any given nucleotide. This\n",
       "error rate defaults to 0.005.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>'illumina4'</code> or <code>'illumina5'</code>: Empirical error models.\n",
       "See <code>?add_platform_error</code> for more information.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>'custom'</code>: A custom error model you've estimated from an\n",
       "RNA-seq data set using <code>GemErr</code>. See <code>?add_platform_error</code>\n",
       "for more info. You will need to provide both <code>model_path</code> and\n",
       "<code>model_prefix</code> if using a custom error model. <code>model_path</code> is\n",
       "the output folder you provided to <code>build_error_model.py</code>. This path\n",
       "should contain either two files suffixed _mate1 and _mate2, or a file\n",
       "suffixed _single. <code>model_prefix</code> is the 'prefix' argument you\n",
       "provided to <code>build_error_model.py</code> and is whatever comes before the\n",
       "_mate1/_mate2 or _single files in <code>model_path</code>.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "</li>\n",
       "<li> <p><code>bias</code> One of 'none', 'rnaf', or 'cdnaf'. 'none'\n",
       "represents uniform fragment selection (every possible fragment in a\n",
       "transcript has equal probability of being in the experiment); 'rnaf'\n",
       "represents positional bias that arises in protocols using RNA\n",
       "fragmentation, and 'cdnaf' represents positional bias arising in protocols\n",
       "that use cDNA fragmentation (Li and Jiang 2012). Using the 'rnaf' model,\n",
       "coverage is higher in the middle of the transcript and lower at both ends,\n",
       "and in the 'cdnaf' model, coverage increases toward the 3' end of the\n",
       "transcript. The probability models used come from Supplementary Figure S3\n",
       "of Li and Jiang (2012). Defaults to 'none' if you don't provide this.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>gcbias</code> list indicating which samples to add GC bias to, and\n",
       "from which models. Should be the same length as <code>sum(num_reps)</code>;\n",
       "entries can be either numeric or of class <code>loess</code>. A numeric entry of\n",
       "0 indicates no GC bias. Numeric entries 1 through 7 correspond to the\n",
       "7 empirical GC models that ship with Polyester, estimated from GEUVADIS\n",
       "HapMap samples NA06985, NA12144, NA12776, NA18858, NA20542, NA20772,\n",
       "and NA20815, respectively. The code used to derive the empirical GC models\n",
       "is available at\n",
       "<a href=\"https://github.com/alyssafrazee/polyester/blob/master/make_gc_bias.R\">https://github.com/alyssafrazee/polyester/blob/master/make_gc_bias.R</a>.\n",
       "A loess entry should be a loess prediction model\n",
       "that takes a GC content percent value (between 0 and 1) a transcript's\n",
       "deviation from overall mean read count based on that GC value. Counts for\n",
       "each replicate will be adjusted based on the GC bias model specified for\n",
       "it. Numeric and loess entries can be mixed. By default, no bias is\n",
       "included.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>frag_GC_bias</code> Either a matrix of dimensions 101 x <code>sum(num_reps)</code>\n",
       "or 'none'. The default is 'none'. If specified, the matrix contains the probabilities\n",
       "(a number in the range [0,1]) that a fragment will appear in the output given its GC content.\n",
       "The first row corresponds to a fragment with GC content of 0 percent, the second row\n",
       "1 percent, the third row 2 percent, etc., and the last row 100 percent.\n",
       "The columns correspond to different probabilites for each sample. Internally,\n",
       "a coin flip (a Bernoulli trial) determines if each fragment is kept, depending on its GC content.\n",
       "Note that the final library size will depend on the elements of the matrix, and\n",
       "it might make sense to scale up the <code>lib_size</code> of the samples with low\n",
       "probabilites in the matrix in the range of the transcriptome GC content distribution.\n",
       "Note that the <code>count_matrix</code> written to <code>outdir</code> contains the counts before\n",
       "applying fragment GC bias.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>strand_specific</code> defaults to <code>FALSE</code>, which means fragments are\n",
       "generated with equal probability from both strands of the transcript sequence.\n",
       "set to <code>TRUE</code> for strand-specific simulation (1st read forward strand,\n",
       "2nd read reverse strand with respect to transcript sequence).\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>meanmodel</code>: set to TRUE if you'd like to set\n",
       "<code>reads_per_transcripts</code> as a function of transcript length. We\n",
       "fit a linear model regressing transcript abundance on transcript length,\n",
       "and setting <code>meanmodel=TRUE</code> means we will use transcript lengths\n",
       "to draw transcript abundance based on that linear model. You can see our\n",
       "modeling code at <a href=\"http://htmlpreview.github.io/?https://github.com/alyssafrazee/polyester_code/blob/master/length_simulation.html\">http://htmlpreview.github.io/?https://github.com/alyssafrazee/polyester_code/blob/master/length_simulation.html</a>\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>write_info</code>: set to FALSE if you do not want files of\n",
       "simulation information written to disk. By default, transcript fold\n",
       "changes and expression status, replicate library sizes and group\n",
       "identifiers, and an R data object of the counts matrix (before\n",
       "application of fragment GC bias) are written to <code>outdir</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>seed</code>: specify a seed (e.g. <code>seed=142</code> or some other\n",
       "integer) to set before randomly drawing read numbers, for reproducibility.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>transcriptid</code>: optional vector of transcript IDs to be written\n",
       "into <code>sim_info.txt</code> and used as transcript identifiers in the output\n",
       "fasta files. Defaults to <code>names(readDNAStringSet(fasta))</code>. This\n",
       "option is useful if default names are very long or contain special\n",
       "characters.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>gzip</code>: pass <code>gzip=TRUE</code> to write gzipped fasta files as\n",
       "output (by default, fasta output files are not compressed when written to\n",
       "disk).\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>exononly</code>: (passed to <code>seq_gtf</code>) if <code>TRUE</code>\n",
       "(as it is by default), only create transcript sequences from the features\n",
       "labeled <code>exon</code> in <code>gtf</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>idfield</code>: (passed to <code>seq_gtf</code>)in the\n",
       "<code>attributes</code> column of <code>gtf</code>, what is the name of the field\n",
       "identifying transcripts? Should be character. Default\n",
       "<code>\"transcript_id\"</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li> <p><code>attrsep</code>: (passed to <code>seq_gtf</code>) in the\n",
       "<code>attributes</code> column of <code>gtf</code>, how are attributes separated?\n",
       "Default <code>\"; \"</code>.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>No return, but simulated reads and a simulation info file are written\n",
       "to <code>outdir</code>. Note that reads are written out transcript by transcript\n",
       "and so need to be shuffled if used as input to quantification algorithms such\n",
       "as eXpress or Salmon.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>'t Hoen PA, et al (2013): Reproducibility of high-throughput mRNA and\n",
       "small RNA sequencing across laboratories. Nature Biotechnology 31(11):\n",
       "1015-1022.\n",
       "</p>\n",
       "<p>Li W and Jiang T (2012): Transcriptome assembly and isoform expression\n",
       "level estimation from biased RNA-Seq reads. Bioinformatics 28(22):\n",
       "2914-2921.\n",
       "</p>\n",
       "<p>McElroy KE, Luciani F and Thomas T (2012): GemSIM: general,\n",
       "error-model based simulator of next-generation sequencing data. BMC\n",
       "Genomics 13(1), 74.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre><code class='language-R'>\n",
       "  ## simulate a few reads from chromosome 22\n",
       "\n",
       "  fastapath = system.file(\"extdata\", \"chr22.fa\", package=\"polyester\")\n",
       "  numtx = count_transcripts(fastapath)\n",
       "  set.seed(4)\n",
       "  fold_change_values = sample(c(0.5, 1, 2), size=2*numtx,\n",
       "     prob=c(0.05, 0.9, 0.05), replace=TRUE)\n",
       "  fold_changes = matrix(fold_change_values, nrow=numtx)\n",
       "  library(Biostrings)\n",
       "  # remove quotes from transcript IDs:\n",
       "  tNames = gsub(\"'\", \"\", names(readDNAStringSet(fastapath)))\n",
       "\n",
       "  simulate_experiment(fastapath, reads_per_transcript=10,\n",
       "     fold_changes=fold_changes, outdir='simulated_reads',\n",
       "     transcriptid=tNames, seed=12)\n",
       "\n",
       "\n",
       "</code></pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>polyester</em> version 1.29.1 ]</div></main>\n",
       "\n",
       "</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{simulate\\_experiment}{simulate RNA-seq experiment using negative binomial model}{simulate.Rul.experiment}\n",
       "%\n",
       "\\begin{Description}\n",
       "create FASTA files containing RNA-seq reads simulated from provided\n",
       "transcripts, with optional differential expression between two groups\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "simulate_experiment(\n",
       "  fasta = NULL,\n",
       "  gtf = NULL,\n",
       "  seqpath = NULL,\n",
       "  outdir = \".\",\n",
       "  num_reps = c(10, 10),\n",
       "  reads_per_transcript = 300,\n",
       "  size = NULL,\n",
       "  fold_changes,\n",
       "  paired = TRUE,\n",
       "  reportCoverage = FALSE,\n",
       "  ...\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{fasta}] path to FASTA file containing transcripts from which to simulate\n",
       "reads. See details.\n",
       "\n",
       "\\item[\\code{gtf}] path to GTF file containing transcript structures from which reads\n",
       "should be simulated. See details.\n",
       "\n",
       "\\item[\\code{seqpath}] path to folder containing one FASTA file (\\code{.fa}\n",
       "extension) for each chromosome in \\code{gtf}. See details.\n",
       "\n",
       "\\item[\\code{outdir}] character, path to folder where simulated reads should be\n",
       "written, with *no* slash at the end. By default, reads are\n",
       "written to current working directory.\n",
       "\n",
       "\\item[\\code{num\\_reps}] How many biological replicates should be in each group? The\n",
       "length \\code{num\\_reps} determines how many groups are in the experiment.\n",
       "For example, \\code{num\\_reps = c(5,6,5)} specifies a 3-group experiment with\n",
       "5 samples in group 1, 6 samples in group 2, and 5 samples in group 3.\n",
       "Defaults to a 2-group experiment with 10 reps per group (i.e.,\n",
       "\\code{c(10,10)}).\n",
       "\n",
       "\\item[\\code{reads\\_per\\_transcript}] baseline mean number of reads to simulate\n",
       "from each transcript. Can be an integer, in which case this many reads\n",
       "are simulated from each transcript, or an integer vector whose length\n",
       "matches the number of transcripts in \\code{fasta}. Default 300. You can\n",
       "also leave \\code{reads\\_per\\_transcript} empty and set \\code{meanmodel=TRUE}\n",
       "to draw baseline mean numbers from a model based on transcript length.\n",
       "\n",
       "\\item[\\code{size}] the negative binomial \\code{size} parameter (see\n",
       "\\code{\\LinkA{NegBinomial}{NegBinomial}}) for the number of reads drawn per transcript.\n",
       "It can be a matrix (where the user can specify the size parameter per\n",
       "transcript, per group), a vector (where the user can specify the size per\n",
       "transcript, perhaps relating to reads\\_per\\_transcript), or a single number,\n",
       "specifying the size for all transcripts and groups.\n",
       "If left NULL, defaults to \\code{reads\\_per\\_transcript * fold\\_changes / 3}.\n",
       "Negative binomial variance is mean + mean\\textasciicircum{}2 / size.\n",
       "\n",
       "\\item[\\code{fold\\_changes}] Matrix specifying multiplicative fold changes\n",
       "between groups. There is no default, so you must provide this argument.\n",
       "In real data sets, lowly-expressed transcripts often show high fold\n",
       "changes between groups, so this can be kept in mind when setting\n",
       "\\code{fold\\_changes} and \\code{reads\\_per\\_transcript}. This argument must\n",
       "have the same number of columns as there are groups as\n",
       "specified by \\code{num\\_reps}, and must have the same number of rows as\n",
       "there are transcripts in \\code{fasta}. A fold change of X in matrix entry\n",
       "i,j means that for replicate j, the baseline mean number of reads\n",
       "(reads\\_per\\_transcript[i]) will be multiplied by X. Note that the\n",
       "multiplication happens before the negative binomial value\n",
       "(for the number of reads that *actually will* be\n",
       "drawn from transcript i, for replicate j) is drawn. This argument is\n",
       "ignored if \\code{length(num\\_reps)} is 1 (meaning you only have 1 group in\n",
       "your simulation).\n",
       "\n",
       "\\item[\\code{paired}] If \\code{TRUE}, paired-end reads are simulated; else\n",
       "single-end reads are simulated. Default \\code{TRUE}\n",
       "\n",
       "\\item[\\code{reportCoverage}] whether to write out coverage information to\n",
       "\\code{sample\\_coverages.rda} file in the \\code{outdir}.\n",
       "defaults to \\code{FALSE}\n",
       "\n",
       "\\item[\\code{...}] any of several other arguments that can be used to add nuance\n",
       "to the simulation. See details.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\n",
       "Reads can either be simulated from a FASTA file of transcripts\n",
       "(provided with the \\code{fasta} argument) or from a GTF file plus DNA\n",
       "sequences (provided with the \\code{gtf} and \\code{seqpath} arguments).\n",
       "Simulating from a GTF file and DNA sequences may be a bit slower: it took\n",
       "about 6 minutes to parse the GTF/sequence files for chromosomes 1-22, X,\n",
       "and Y in hg19.\n",
       "\n",
       "Several optional parameters can be passed to this function to adjust the\n",
       "simulation. The options are:\n",
       "\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} \\code{readlen}: read length. Default 100.\n",
       "\\item{} \\code{lib\\_sizes}: Library size factors for the biological replicates.\n",
       "\\code{lib\\_sizes} should have length equal to the total number of\n",
       "replicates in the experiment, i.e., \\code{sum(num\\_reps)}. For each\n",
       "replicate, once the number of reads to simulate from each transcript for\n",
       "that replicate is known, all read numbers across all transcripts from that\n",
       "replicate are multiplied by the corresponding entry in \\code{lib\\_sizes}.\n",
       "\\item{} \\code{distr} One of 'normal', 'empirical', or 'custom', which\n",
       "specifies the distribution from which to draw RNA fragment lengths. If\n",
       "'normal', draw fragment lengths from a normal distribution. You can provide\n",
       "the mean of that normal distribution with \\code{fraglen} (defaults to 250)\n",
       "and the standard deviation of that normal distribution with \\code{fragsd}\n",
       "(defaults to 25). You can provide a single number for each, or a\n",
       "vector with length equal to the total number of samples.\n",
       "If 'empirical', draw fragment lengths\n",
       "from a fragment length distribution estimated from a real data set. If\n",
       "'custom', draw fragment lengths from a custom distribution, which you can\n",
       "provide as the \\code{custdens} argument. \\code{custdens} should be a\n",
       "density fitted using \\code{\\LinkA{logspline}{logspline}}.\n",
       "\\item{} \\code{error\\_model}: The error model can be one of:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} \\code{'uniform'}: errors are distributed uniformly across reads.\n",
       "You can also provide an \\code{'error\\_rate'} parameter, giving the overall\n",
       "probability of making a sequencing error at any given nucleotide. This\n",
       "error rate defaults to 0.005.\n",
       "\\item{} \\code{'illumina4'} or \\code{'illumina5'}: Empirical error models.\n",
       "See \\code{?add\\_platform\\_error} for more information.\n",
       "\\item{} \\code{'custom'}: A custom error model you've estimated from an\n",
       "RNA-seq data set using \\code{GemErr}. See \\code{?add\\_platform\\_error}\n",
       "for more info. You will need to provide both \\code{model\\_path} and\n",
       "\\code{model\\_prefix} if using a custom error model. \\code{model\\_path} is\n",
       "the output folder you provided to \\code{build\\_error\\_model.py}. This path\n",
       "should contain either two files suffixed \\_mate1 and \\_mate2, or a file\n",
       "suffixed \\_single. \\code{model\\_prefix} is the 'prefix' argument you\n",
       "provided to \\code{build\\_error\\_model.py} and is whatever comes before the\n",
       "\\_mate1/\\_mate2 or \\_single files in \\code{model\\_path}.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\item{} \\code{bias} One of 'none', 'rnaf', or 'cdnaf'. 'none'\n",
       "represents uniform fragment selection (every possible fragment in a\n",
       "transcript has equal probability of being in the experiment); 'rnaf'\n",
       "represents positional bias that arises in protocols using RNA\n",
       "fragmentation, and 'cdnaf' represents positional bias arising in protocols\n",
       "that use cDNA fragmentation (Li and Jiang 2012). Using the 'rnaf' model,\n",
       "coverage is higher in the middle of the transcript and lower at both ends,\n",
       "and in the 'cdnaf' model, coverage increases toward the 3' end of the\n",
       "transcript. The probability models used come from Supplementary Figure S3\n",
       "of Li and Jiang (2012). Defaults to 'none' if you don't provide this.\n",
       "\\item{} \\code{gcbias} list indicating which samples to add GC bias to, and\n",
       "from which models. Should be the same length as \\code{sum(num\\_reps)};\n",
       "entries can be either numeric or of class \\code{loess}. A numeric entry of\n",
       "0 indicates no GC bias. Numeric entries 1 through 7 correspond to the\n",
       "7 empirical GC models that ship with Polyester, estimated from GEUVADIS\n",
       "HapMap samples NA06985, NA12144, NA12776, NA18858, NA20542, NA20772,\n",
       "and NA20815, respectively. The code used to derive the empirical GC models\n",
       "is available at\n",
       "\\url{https://github.com/alyssafrazee/polyester/blob/master/make_gc_bias.R}.\n",
       "A loess entry should be a loess prediction model\n",
       "that takes a GC content percent value (between 0 and 1) a transcript's\n",
       "deviation from overall mean read count based on that GC value. Counts for\n",
       "each replicate will be adjusted based on the GC bias model specified for\n",
       "it. Numeric and loess entries can be mixed. By default, no bias is\n",
       "included.\n",
       "\\item{} \\code{frag\\_GC\\_bias} Either a matrix of dimensions 101 x \\code{sum(num\\_reps)}\n",
       "or 'none'. The default is 'none'. If specified, the matrix contains the probabilities\n",
       "(a number in the range [0,1]) that a fragment will appear in the output given its GC content.\n",
       "The first row corresponds to a fragment with GC content of 0 percent, the second row\n",
       "1 percent, the third row 2 percent, etc., and the last row 100 percent.\n",
       "The columns correspond to different probabilites for each sample. Internally,\n",
       "a coin flip (a Bernoulli trial) determines if each fragment is kept, depending on its GC content.\n",
       "Note that the final library size will depend on the elements of the matrix, and\n",
       "it might make sense to scale up the \\code{lib\\_size} of the samples with low\n",
       "probabilites in the matrix in the range of the transcriptome GC content distribution.\n",
       "Note that the \\code{count\\_matrix} written to \\code{outdir} contains the counts before\n",
       "applying fragment GC bias.\n",
       "\\item{} \\code{strand\\_specific} defaults to \\code{FALSE}, which means fragments are\n",
       "generated with equal probability from both strands of the transcript sequence.\n",
       "set to \\code{TRUE} for strand-specific simulation (1st read forward strand,\n",
       "2nd read reverse strand with respect to transcript sequence).\n",
       "\\item{} \\code{meanmodel}: set to TRUE if you'd like to set\n",
       "\\code{reads\\_per\\_transcripts} as a function of transcript length. We\n",
       "fit a linear model regressing transcript abundance on transcript length,\n",
       "and setting \\code{meanmodel=TRUE} means we will use transcript lengths\n",
       "to draw transcript abundance based on that linear model. You can see our\n",
       "modeling code at \\url{http://htmlpreview.github.io/?https://github.com/alyssafrazee/polyester_code/blob/master/length_simulation.html}\n",
       "\\item{} \\code{write\\_info}: set to FALSE if you do not want files of\n",
       "simulation information written to disk. By default, transcript fold\n",
       "changes and expression status, replicate library sizes and group\n",
       "identifiers, and an R data object of the counts matrix (before\n",
       "application of fragment GC bias) are written to \\code{outdir}.\n",
       "\\item{} \\code{seed}: specify a seed (e.g. \\code{seed=142} or some other\n",
       "integer) to set before randomly drawing read numbers, for reproducibility.\n",
       "\\item{} \\code{transcriptid}: optional vector of transcript IDs to be written\n",
       "into \\code{sim\\_info.txt} and used as transcript identifiers in the output\n",
       "fasta files. Defaults to \\code{names(readDNAStringSet(fasta))}. This\n",
       "option is useful if default names are very long or contain special\n",
       "characters.\n",
       "\\item{} \\code{gzip}: pass \\code{gzip=TRUE} to write gzipped fasta files as\n",
       "output (by default, fasta output files are not compressed when written to\n",
       "disk).\n",
       "\\item{} \\code{exononly}: (passed to \\code{\\LinkA{seq\\_gtf}{seq.Rul.gtf}}) if \\code{TRUE}\n",
       "(as it is by default), only create transcript sequences from the features\n",
       "labeled \\code{exon} in \\code{gtf}.\n",
       "\\item{} \\code{idfield}: (passed to \\code{\\LinkA{seq\\_gtf}{seq.Rul.gtf}})in the\n",
       "\\code{attributes} column of \\code{gtf}, what is the name of the field\n",
       "identifying transcripts? Should be character. Default\n",
       "\\code{\"transcript\\_id\"}.\n",
       "\\item{} \\code{attrsep}: (passed to \\code{\\LinkA{seq\\_gtf}{seq.Rul.gtf}}) in the\n",
       "\\code{attributes} column of \\code{gtf}, how are attributes separated?\n",
       "Default \\code{\"; \"}.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "No return, but simulated reads and a simulation info file are written\n",
       "to \\code{outdir}. Note that reads are written out transcript by transcript\n",
       "and so need to be shuffled if used as input to quantification algorithms such\n",
       "as eXpress or Salmon.\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{References}\n",
       "'t Hoen PA, et al (2013): Reproducibility of high-throughput mRNA and\n",
       "small RNA sequencing across laboratories. Nature Biotechnology 31(11):\n",
       "1015-1022.\n",
       "\n",
       "Li W and Jiang T (2012): Transcriptome assembly and isoform expression\n",
       "level estimation from biased RNA-Seq reads. Bioinformatics 28(22):\n",
       "2914-2921.\n",
       "\n",
       "McElroy KE, Luciani F and Thomas T (2012): GemSIM: general,\n",
       "error-model based simulator of next-generation sequencing data. BMC\n",
       "Genomics 13(1), 74.\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "\n",
       "  ## simulate a few reads from chromosome 22\n",
       "\n",
       "  fastapath = system.file(\"extdata\", \"chr22.fa\", package=\"polyester\")\n",
       "  numtx = count_transcripts(fastapath)\n",
       "  set.seed(4)\n",
       "  fold_change_values = sample(c(0.5, 1, 2), size=2*numtx,\n",
       "     prob=c(0.05, 0.9, 0.05), replace=TRUE)\n",
       "  fold_changes = matrix(fold_change_values, nrow=numtx)\n",
       "  library(Biostrings)\n",
       "  # remove quotes from transcript IDs:\n",
       "  tNames = gsub(\"'\", \"\", names(readDNAStringSet(fastapath)))\n",
       "\n",
       "  simulate_experiment(fastapath, reads_per_transcript=10,\n",
       "     fold_changes=fold_changes, outdir='simulated_reads',\n",
       "     transcriptid=tNames, seed=12)\n",
       "\n",
       "\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "simulate_experiment         package:polyester          R Documentation\n",
       "\n",
       "_\bs_\bi_\bm_\bu_\bl_\ba_\bt_\be _\bR_\bN_\bA-_\bs_\be_\bq _\be_\bx_\bp_\be_\br_\bi_\bm_\be_\bn_\bt _\bu_\bs_\bi_\bn_\bg _\bn_\be_\bg_\ba_\bt_\bi_\bv_\be _\bb_\bi_\bn_\bo_\bm_\bi_\ba_\bl _\bm_\bo_\bd_\be_\bl\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     create FASTA files containing RNA-seq reads simulated from\n",
       "     provided transcripts, with optional differential expression\n",
       "     between two groups\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     simulate_experiment(\n",
       "       fasta = NULL,\n",
       "       gtf = NULL,\n",
       "       seqpath = NULL,\n",
       "       outdir = \".\",\n",
       "       num_reps = c(10, 10),\n",
       "       reads_per_transcript = 300,\n",
       "       size = NULL,\n",
       "       fold_changes,\n",
       "       paired = TRUE,\n",
       "       reportCoverage = FALSE,\n",
       "       ...\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "   fasta: path to FASTA file containing transcripts from which to\n",
       "          simulate reads. See details.\n",
       "\n",
       "     gtf: path to GTF file containing transcript structures from which\n",
       "          reads should be simulated. See details.\n",
       "\n",
       " seqpath: path to folder containing one FASTA file (‘.fa’ extension)\n",
       "          for each chromosome in ‘gtf’. See details.\n",
       "\n",
       "  outdir: character, path to folder where simulated reads should be\n",
       "          written, with *no* slash at the end. By default, reads are\n",
       "          written to current working directory.\n",
       "\n",
       "num_reps: How many biological replicates should be in each group? The\n",
       "          length ‘num_reps’ determines how many groups are in the\n",
       "          experiment. For example, ‘num_reps = c(5,6,5)’ specifies a\n",
       "          3-group experiment with 5 samples in group 1, 6 samples in\n",
       "          group 2, and 5 samples in group 3. Defaults to a 2-group\n",
       "          experiment with 10 reps per group (i.e., ‘c(10,10)’).\n",
       "\n",
       "reads_per_transcript: baseline mean number of reads to simulate from\n",
       "          each transcript. Can be an integer, in which case this many\n",
       "          reads are simulated from each transcript, or an integer\n",
       "          vector whose length matches the number of transcripts in\n",
       "          ‘fasta’. Default 300. You can also leave\n",
       "          ‘reads_per_transcript’ empty and set ‘meanmodel=TRUE’ to draw\n",
       "          baseline mean numbers from a model based on transcript\n",
       "          length.\n",
       "\n",
       "    size: the negative binomial ‘size’ parameter (see ‘NegBinomial’)\n",
       "          for the number of reads drawn per transcript. It can be a\n",
       "          matrix (where the user can specify the size parameter per\n",
       "          transcript, per group), a vector (where the user can specify\n",
       "          the size per transcript, perhaps relating to\n",
       "          reads_per_transcript), or a single number, specifying the\n",
       "          size for all transcripts and groups. If left NULL, defaults\n",
       "          to ‘reads_per_transcript * fold_changes / 3’. Negative\n",
       "          binomial variance is mean + mean^2 / size.\n",
       "\n",
       "fold_changes: Matrix specifying multiplicative fold changes between\n",
       "          groups. There is no default, so you must provide this\n",
       "          argument. In real data sets, lowly-expressed transcripts\n",
       "          often show high fold changes between groups, so this can be\n",
       "          kept in mind when setting ‘fold_changes’ and\n",
       "          ‘reads_per_transcript’. This argument must have the same\n",
       "          number of columns as there are groups as specified by\n",
       "          ‘num_reps’, and must have the same number of rows as there\n",
       "          are transcripts in ‘fasta’. A fold change of X in matrix\n",
       "          entry i,j means that for replicate j, the baseline mean\n",
       "          number of reads (reads_per_transcript[i]) will be multiplied\n",
       "          by X. Note that the multiplication happens before the\n",
       "          negative binomial value (for the number of reads that\n",
       "          *actually will* be drawn from transcript i, for replicate j)\n",
       "          is drawn. This argument is ignored if ‘length(num_reps)’ is 1\n",
       "          (meaning you only have 1 group in your simulation).\n",
       "\n",
       "  paired: If ‘TRUE’, paired-end reads are simulated; else single-end\n",
       "          reads are simulated. Default ‘TRUE’\n",
       "\n",
       "reportCoverage: whether to write out coverage information to\n",
       "          ‘sample_coverages.rda’ file in the ‘outdir’. defaults to\n",
       "          ‘FALSE’\n",
       "\n",
       "     ...: any of several other arguments that can be used to add nuance\n",
       "          to the simulation. See details.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     Reads can either be simulated from a FASTA file of transcripts\n",
       "     (provided with the ‘fasta’ argument) or from a GTF file plus DNA\n",
       "     sequences (provided with the ‘gtf’ and ‘seqpath’ arguments).\n",
       "     Simulating from a GTF file and DNA sequences may be a bit slower:\n",
       "     it took about 6 minutes to parse the GTF/sequence files for\n",
       "     chromosomes 1-22, X, and Y in hg19.\n",
       "\n",
       "     Several optional parameters can be passed to this function to\n",
       "     adjust the simulation. The options are:\n",
       "\n",
       "        • ‘readlen’: read length. Default 100.\n",
       "\n",
       "        • ‘lib_sizes’: Library size factors for the biological\n",
       "          replicates.  ‘lib_sizes’ should have length equal to the\n",
       "          total number of replicates in the experiment, i.e.,\n",
       "          ‘sum(num_reps)’. For each replicate, once the number of reads\n",
       "          to simulate from each transcript for that replicate is known,\n",
       "          all read numbers across all transcripts from that replicate\n",
       "          are multiplied by the corresponding entry in ‘lib_sizes’.\n",
       "\n",
       "        • ‘distr’ One of 'normal', 'empirical', or 'custom', which\n",
       "          specifies the distribution from which to draw RNA fragment\n",
       "          lengths. If 'normal', draw fragment lengths from a normal\n",
       "          distribution. You can provide the mean of that normal\n",
       "          distribution with ‘fraglen’ (defaults to 250) and the\n",
       "          standard deviation of that normal distribution with ‘fragsd’\n",
       "          (defaults to 25). You can provide a single number for each,\n",
       "          or a vector with length equal to the total number of samples.\n",
       "          If 'empirical', draw fragment lengths from a fragment length\n",
       "          distribution estimated from a real data set. If 'custom',\n",
       "          draw fragment lengths from a custom distribution, which you\n",
       "          can provide as the ‘custdens’ argument. ‘custdens’ should be\n",
       "          a density fitted using ‘logspline’.\n",
       "\n",
       "        • ‘error_model’: The error model can be one of:\n",
       "\n",
       "            • ‘'uniform'’: errors are distributed uniformly across\n",
       "              reads.  You can also provide an ‘'error_rate'’ parameter,\n",
       "              giving the overall probability of making a sequencing\n",
       "              error at any given nucleotide. This error rate defaults\n",
       "              to 0.005.\n",
       "\n",
       "            • ‘'illumina4'’ or ‘'illumina5'’: Empirical error models.\n",
       "              See ‘?add_platform_error’ for more information.\n",
       "\n",
       "            • ‘'custom'’: A custom error model you've estimated from an\n",
       "              RNA-seq data set using ‘GemErr’. See\n",
       "              ‘?add_platform_error’ for more info. You will need to\n",
       "              provide both ‘model_path’ and ‘model_prefix’ if using a\n",
       "              custom error model. ‘model_path’ is the output folder you\n",
       "              provided to ‘build_error_model.py’. This path should\n",
       "              contain either two files suffixed _mate1 and _mate2, or a\n",
       "              file suffixed _single. ‘model_prefix’ is the 'prefix'\n",
       "              argument you provided to ‘build_error_model.py’ and is\n",
       "              whatever comes before the _mate1/_mate2 or _single files\n",
       "              in ‘model_path’.\n",
       "\n",
       "        • ‘bias’ One of 'none', 'rnaf', or 'cdnaf'. 'none' represents\n",
       "          uniform fragment selection (every possible fragment in a\n",
       "          transcript has equal probability of being in the experiment);\n",
       "          'rnaf' represents positional bias that arises in protocols\n",
       "          using RNA fragmentation, and 'cdnaf' represents positional\n",
       "          bias arising in protocols that use cDNA fragmentation (Li and\n",
       "          Jiang 2012). Using the 'rnaf' model, coverage is higher in\n",
       "          the middle of the transcript and lower at both ends, and in\n",
       "          the 'cdnaf' model, coverage increases toward the 3' end of\n",
       "          the transcript. The probability models used come from\n",
       "          Supplementary Figure S3 of Li and Jiang (2012). Defaults to\n",
       "          'none' if you don't provide this.\n",
       "\n",
       "        • ‘gcbias’ list indicating which samples to add GC bias to, and\n",
       "          from which models. Should be the same length as\n",
       "          ‘sum(num_reps)’; entries can be either numeric or of class\n",
       "          ‘loess’. A numeric entry of 0 indicates no GC bias. Numeric\n",
       "          entries 1 through 7 correspond to the 7 empirical GC models\n",
       "          that ship with Polyester, estimated from GEUVADIS HapMap\n",
       "          samples NA06985, NA12144, NA12776, NA18858, NA20542, NA20772,\n",
       "          and NA20815, respectively. The code used to derive the\n",
       "          empirical GC models is available at\n",
       "          <https://github.com/alyssafrazee/polyester/blob/master/make_gc_bias.R>.\n",
       "          A loess entry should be a loess prediction model that takes a\n",
       "          GC content percent value (between 0 and 1) a transcript's\n",
       "          deviation from overall mean read count based on that GC\n",
       "          value. Counts for each replicate will be adjusted based on\n",
       "          the GC bias model specified for it. Numeric and loess entries\n",
       "          can be mixed. By default, no bias is included.\n",
       "\n",
       "        • ‘frag_GC_bias’ Either a matrix of dimensions 101 x\n",
       "          ‘sum(num_reps)’ or 'none'. The default is 'none'. If\n",
       "          specified, the matrix contains the probabilities (a number in\n",
       "          the range [0,1]) that a fragment will appear in the output\n",
       "          given its GC content.  The first row corresponds to a\n",
       "          fragment with GC content of 0 percent, the second row 1\n",
       "          percent, the third row 2 percent, etc., and the last row 100\n",
       "          percent.  The columns correspond to different probabilites\n",
       "          for each sample. Internally, a coin flip (a Bernoulli trial)\n",
       "          determines if each fragment is kept, depending on its GC\n",
       "          content.  Note that the final library size will depend on the\n",
       "          elements of the matrix, and it might make sense to scale up\n",
       "          the ‘lib_size’ of the samples with low probabilites in the\n",
       "          matrix in the range of the transcriptome GC content\n",
       "          distribution.  Note that the ‘count_matrix’ written to\n",
       "          ‘outdir’ contains the counts before applying fragment GC\n",
       "          bias.\n",
       "\n",
       "        • ‘strand_specific’ defaults to ‘FALSE’, which means fragments\n",
       "          are generated with equal probability from both strands of the\n",
       "          transcript sequence.  set to ‘TRUE’ for strand-specific\n",
       "          simulation (1st read forward strand, 2nd read reverse strand\n",
       "          with respect to transcript sequence).\n",
       "\n",
       "        • ‘meanmodel’: set to TRUE if you'd like to set\n",
       "          ‘reads_per_transcripts’ as a function of transcript length.\n",
       "          We fit a linear model regressing transcript abundance on\n",
       "          transcript length, and setting ‘meanmodel=TRUE’ means we will\n",
       "          use transcript lengths to draw transcript abundance based on\n",
       "          that linear model. You can see our modeling code at\n",
       "          <http://htmlpreview.github.io/?https://github.com/alyssafrazee/polyester_code/blob/master/length_simulation.html>\n",
       "\n",
       "        • ‘write_info’: set to FALSE if you do not want files of\n",
       "          simulation information written to disk. By default,\n",
       "          transcript fold changes and expression status, replicate\n",
       "          library sizes and group identifiers, and an R data object of\n",
       "          the counts matrix (before application of fragment GC bias)\n",
       "          are written to ‘outdir’.\n",
       "\n",
       "        • ‘seed’: specify a seed (e.g. ‘seed=142’ or some other\n",
       "          integer) to set before randomly drawing read numbers, for\n",
       "          reproducibility.\n",
       "\n",
       "        • ‘transcriptid’: optional vector of transcript IDs to be\n",
       "          written into ‘sim_info.txt’ and used as transcript\n",
       "          identifiers in the output fasta files. Defaults to\n",
       "          ‘names(readDNAStringSet(fasta))’. This option is useful if\n",
       "          default names are very long or contain special characters.\n",
       "\n",
       "        • ‘gzip’: pass ‘gzip=TRUE’ to write gzipped fasta files as\n",
       "          output (by default, fasta output files are not compressed\n",
       "          when written to disk).\n",
       "\n",
       "        • ‘exononly’: (passed to ‘seq_gtf’) if ‘TRUE’ (as it is by\n",
       "          default), only create transcript sequences from the features\n",
       "          labeled ‘exon’ in ‘gtf’.\n",
       "\n",
       "        • ‘idfield’: (passed to ‘seq_gtf’)in the ‘attributes’ column of\n",
       "          ‘gtf’, what is the name of the field identifying transcripts?\n",
       "          Should be character. Default ‘\"transcript_id\"’.\n",
       "\n",
       "        • ‘attrsep’: (passed to ‘seq_gtf’) in the ‘attributes’ column\n",
       "          of ‘gtf’, how are attributes separated?  Default ‘\"; \"’.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     No return, but simulated reads and a simulation info file are\n",
       "     written to ‘outdir’. Note that reads are written out transcript by\n",
       "     transcript and so need to be shuffled if used as input to\n",
       "     quantification algorithms such as eXpress or Salmon.\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     't Hoen PA, et al (2013): Reproducibility of high-throughput mRNA\n",
       "     and small RNA sequencing across laboratories. Nature Biotechnology\n",
       "     31(11): 1015-1022.\n",
       "\n",
       "     Li W and Jiang T (2012): Transcriptome assembly and isoform\n",
       "     expression level estimation from biased RNA-Seq reads.\n",
       "     Bioinformatics 28(22): 2914-2921.\n",
       "\n",
       "     McElroy KE, Luciani F and Thomas T (2012): GemSIM: general,\n",
       "     error-model based simulator of next-generation sequencing data.\n",
       "     BMC Genomics 13(1), 74.\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "       ## simulate a few reads from chromosome 22\n",
       "     \n",
       "       fastapath = system.file(\"extdata\", \"chr22.fa\", package=\"polyester\")\n",
       "       numtx = count_transcripts(fastapath)\n",
       "       set.seed(4)\n",
       "       fold_change_values = sample(c(0.5, 1, 2), size=2*numtx,\n",
       "          prob=c(0.05, 0.9, 0.05), replace=TRUE)\n",
       "       fold_changes = matrix(fold_change_values, nrow=numtx)\n",
       "       library(Biostrings)\n",
       "       # remove quotes from transcript IDs:\n",
       "       tNames = gsub(\"'\", \"\", names(readDNAStringSet(fastapath)))\n",
       "     \n",
       "       simulate_experiment(fastapath, reads_per_transcript=10,\n",
       "          fold_changes=fold_changes, outdir='simulated_reads',\n",
       "          transcriptid=tNames, seed=12)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?simulate_experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc6d0b00-b1b4-4779-a2d7-9591b338dd96",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
